pytorch_lightning/utilities/auto_restart.py

```diff
@@ -207,29 +207,189 @@ def __iter__(self) -> Iterator:
         return self
 
     def __next__(self) -> Dict[str, Any]:
-        # fetch next data
-        data = next(self.iter_data)
+        return next(self.iter_data)
 
-        # create current samplers state_dict
-        worker_info = get_worker_info()
-        state_dicts = {"id": worker_info.id if worker_info is not None else 0}
-        state_dicts.update({k: v.state_dict() for k, v in self.samplers.items()})
+    @staticmethod
+    def store_samplers_state_dict(iterator: Iterator, sampler_state_dict: List) -> None:
+        iterator_state_dict = getattr(iterator, "_sampler_state_dict", None)
+        iterator_state_dict_cache = getattr(iterator, "_sampler_state_dict_cache", None)
+        # handle the logic this way due Trainer prefecting.
+        if not iterator_state_dict:
+            iterator._sampler_state_dict = sampler_state_dict
+        else:
+            if iterator_state_dict_cache is None:
+                iterator._sampler_state_dict_cache = sampler_state_dict
+            else:
+                for attr_cache, state_dict in zip(iterator_state_dict, iterator._sampler_state_dict_cache):
+                    for k, v in state_dict.items():
+                        attr_cache[k].update(v)
+                iterator._sampler_state_dict_cache = sampler_state_dict
 
-        # return both current data and samplers ``state_dict``.
-        return {"data": data, AutoRestartBatchKeys.PL_SAMPLERS: state_dicts}
+    @staticmethod
+    def _sanitize_batch_from_sampler_state(data: Any, state_dicts: List):
+        """
+        This function is used to remove the sampler state dict from provided data batch.
+        The custom data has this format:
+
+        .. code-block:: python
+
+            {
+                "batch": data returned by DataLoader
+                "__pl_samplers": {
```

Comment:
```suggestion
                "__pl_samplers__": {
```

Still think it should be `pl_samplers`. Double underscores are meant to be reserved for Python's dunder methods

Reply:
@carmocca double underscores at the beginning and end are meant to be researved for that. Double underscores at the beginning are only indicating private attributes (vs. the protected ones with only one underscore) which is why I think it should stay like this.

Reply:
I thought you were trying to use the same name as

```python
pytorch_lightning//utilities/enums.py:121:    PL_SAMPLERS = "__pl_samplers__"
pytorch_lightning//utilities/auto_restart.py:386:            "__pl_samplers__": {
```

Reply:
Ok, that's my bad then actually :D Then I think this should always be `__pl_samplers`

Reply:
Note that double underscore is overkill here as the conversion from `__{name}` to `_{class}_{name}` does not get done

Reply:
True, But personally I'd still stick with that pattern if you don't have any objections